ambry-network/src/main/java/com.github.ambry.network/SocketNetworkClient.java

```diff
@@ -0,0 +1,416 @@
+  private final HashMap<String, RequestMetadata> pendingConnectionsToAssociatedRequests;
+  private final AtomicLong numPendingRequests;
+  private final int checkoutTimeoutMs;
+  private boolean closed = false;
+  private static final Logger logger = LoggerFactory.getLogger(SocketNetworkClient.class);
+
+  /**
+   * Instantiates a SocketNetworkClient.
+   * @param selector the {@link Selector} for this SocketNetworkClient
+   * @param maxConnectionsPerPortPlainText the maximum number of connections per node per plain text port
+   * @param maxConnectionsPerPortSsl the maximum number of connections per node per ssl port
+   * @param networkConfig the {@link NetworkConfig} for this SocketNetworkClient
+   * @param networkMetrics the metrics to track the network related metrics
+   * @param checkoutTimeoutMs the maximum time a request should remain in this SocketNetworkClient's pending queue waiting
+   *                          for an available connection to its destination.
+   * @param time The Time instance to use.
+   */
+  public SocketNetworkClient(Selector selector, NetworkConfig networkConfig, NetworkMetrics networkMetrics,
+      int maxConnectionsPerPortPlainText, int maxConnectionsPerPortSsl, int checkoutTimeoutMs, Time time) {
+    this.selector = selector;
+    this.connectionTracker = new ConnectionTracker(maxConnectionsPerPortPlainText, maxConnectionsPerPortSsl);
+    this.networkConfig = networkConfig;
+    this.networkMetrics = networkMetrics;
+    this.checkoutTimeoutMs = checkoutTimeoutMs;
+    this.time = time;
+    pendingRequests = new LinkedList<>();
+    numPendingRequests = new AtomicLong(0);
+    connectionIdToRequestInFlight = new HashMap<>();
+    correlationIdInFlightToConnectionId = new HashMap<>();
+    pendingConnectionsToAssociatedRequests = new HashMap<>();
+    networkMetrics.registerNetworkClientPendingConnections(numPendingRequests);
+  }
+
+  /**
+   * Attempt to send the given requests and poll for responses from the network via the associated selector. Any
+   * requests that could not be sent out will be added to a queue. Every time this method is called, it will first
+   * attempt sending the requests in the queue (or time them out) and then attempt sending the newly added requests.
+   * @param requestsToSend the list of {@link RequestInfo} representing the requests that need to be sent out. This
+   *                     could be empty.
+   * @param requestsToDrop the list of correlation IDs representing the requests that can be dropped by
+   *                       closing the connection.
+   * @param pollTimeoutMs the poll timeout.
+   * @return a list of {@link ResponseInfo} representing the responses received for any requests that were sent out
+   * so far.
+   * @throws IllegalStateException if the SocketNetworkClient is closed.
+   */
+  public List<ResponseInfo> sendAndPoll(List<RequestInfo> requestsToSend, Set<Integer> requestsToDrop,
```

Comment:
add `@Override` annotations for all the methods that are defined in the interface

Reply:
ok